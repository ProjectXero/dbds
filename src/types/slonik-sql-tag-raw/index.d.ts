declare module 'slonik-sql-tag-raw' {
  import type { ValueExpressionType, SqlSqlTokenType } from 'slonik'

  export function raw(
    sql: string,
    values?: ReadonlyArray<ValueExpressionType>
  ): SqlSqlTokenType
}
