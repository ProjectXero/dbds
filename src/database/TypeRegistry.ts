import { factory, TypeNode } from 'typescript'

export const UNKNOWN = factory.createTypeReferenceNode('unknown')

export const defaultTypeMap = Object.freeze({
  bpchar: 'string',
  char: 'string',
  varchar: 'string',
  text: 'string',
  citext: 'string',
  uuid: 'string',
  bytea: 'string',
  inet: 'string',
  time: 'string',
  timetz: 'string',
  interval: 'string',
  name: 'string',
  int2: 'number',
  int4: 'number',
  int8: 'number',
  float4: 'number',
  float8: 'number',
  numeric: 'number',
  money: 'number',
  oid: 'number',
  bool: 'boolean',
  json: 'unknown',
  jsonb: 'unknown',
  date: 'Date',
  timestamp: 'Date',
  timestamptz: 'Date',
})

export default class TypeRegistry {
  public readonly typeMap: Record<string, string>

  constructor() {
    this.typeMap = { ...defaultTypeMap }
  }

  public isKnownType(typeName: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.typeMap, typeName)
  }

  public getTypeNode(typeName: string): TypeNode {
    if (this.isKnownType(typeName)) {
      return factory.createTypeReferenceNode(this.typeMap[typeName])
    }

    console.warn(`Unknown type detected: '${typeName}'. This is probably a bug.`)
    this.registerType(typeName, 'unknown')

    return UNKNOWN
  }

  public registerType(typeName: string, target: string): void {
    if (this.isKnownType(typeName)) {
      console.warn(`Re-registering known type '${typeName}': ${this.typeMap[typeName]} => ${target}`)
    }

    this.typeMap[typeName] = target;
  }
}
