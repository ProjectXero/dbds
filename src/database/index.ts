export { default as SchemaInfo } from './SchemaInfo'
export * from './SchemaInfo'
export { default as TypeMapper } from './TypeMapper'
