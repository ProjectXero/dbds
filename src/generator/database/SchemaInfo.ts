import { DatabasePoolType, sql, TaggedTemplateLiteralInvocationType } from 'slonik'

export interface EnumInfo {
  name: string
  values: string[]
}

export interface ColumnInfo {
  name: string
  nullable: boolean
  hasDefault: boolean
  order: number
  type: string
  isArray: boolean
}

export interface TableInfo {
  name: string
  canInsert: boolean
  columns: readonly ColumnInfo[]
}

export default class SchemaInfo {
  constructor(
    private readonly pool: DatabasePoolType,
    public readonly name: string
  ) { }

  public tableQuery(): TaggedTemplateLiteralInvocationType<Omit<TableInfo, 'columns'>> {
    return sql<Omit<TableInfo, 'columns'>>`
      SELECT table_name AS "name"
          , (is_insertable_into = 'YES') AS "canInsert"
          , pg_catalog.obj_description(('"' || table_name || '"')::regclass::oid) AS "comment"
      FROM information_schema.tables
      WHERE table_schema = ${this.name}
      ORDER BY table_name ASC
    `
  }

  public columnQuery(table: string): TaggedTemplateLiteralInvocationType<ColumnInfo> {
    return sql<ColumnInfo>`
      SELECT c.column_name AS "name"
           , (c.is_nullable = 'YES') AS "nullable"
           , (c.column_default IS NOT NULL) AS "hasDefault"
           , c.ordinal_position AS "order"
           , COALESCE(de.udt_name, dc.udt_name, e.udt_name, e.data_type, c.udt_name, c.data_type) AS "type"
           , (c.data_type = 'ARRAY') AS "isArray"
           , pg_catalog.col_description(('"' || c.table_name || '"')::regclass::oid, c.ordinal_position::int) AS "comment"
      FROM
        information_schema.columns c LEFT JOIN
        information_schema.element_types e ON (
          c.data_type = 'ARRAY' AND
          c.table_schema = e.object_schema AND
          c.table_name = e.object_name AND
          c.dtd_identifier = e.collection_type_identifier
        ) LEFT JOIN
        information_schema.domains dc ON (
          c.data_type = 'USER-DEFINED' AND
          c.table_schema = dc.domain_schema AND
          c.domain_name = dc.domain_name
        ) LEFT JOIN
        information_schema.domains de ON (
          e.data_type = 'USER-DEFINED' AND
          e.object_schema = de.domain_schema AND
          e.udt_name = de.domain_name
        )
      WHERE (
        c.table_schema = ${this.name} AND
        c.table_name = ${table}
      )
      ORDER BY c.ordinal_position
    `
  }

  public enumQuery(): TaggedTemplateLiteralInvocationType<EnumInfo> {
    return sql<EnumInfo>`
      SELECT t.typname AS "name"
           , array_agg(e.enumlabel)::TEXT[] AS "values"
           , pg_catalog.obj_description(e.enumtypid) AS "comment"
      FROM
        pg_type t JOIN
        pg_enum e ON (
          t.oid = e.enumtypid
        ) JOIN
        pg_catalog.pg_namespace n ON (
          n.oid = t.typnamespace
        )
      WHERE (
        n.nspname = ${this.name}
      )
      GROUP BY n.nspname
             , t.typname
             , e.enumtypid
      ORDER BY t.typname
    `
  }

  public async disconnect(): Promise<void> {
    await this.pool.end()
  }

  public async getTables(): Promise<readonly TableInfo[]> {
    const tables = await this.pool.any(this.tableQuery())

    return Promise.all(
      tables.map<Promise<TableInfo>>(async (table) => {
        return {
          ...table,
          columns: await this.getTableColumns(table.name)
        }
      })
    )
  }

  public async getTableColumns(table: string): Promise<readonly ColumnInfo[]> {
    return await this.pool.many(this.columnQuery(table))
  }

  public async getEnums(): Promise<readonly EnumInfo[]> {
    return await this.pool.any(this.enumQuery())
  }

}
