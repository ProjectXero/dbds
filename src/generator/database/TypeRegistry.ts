import { factory, TypeNode } from 'typescript'

export const UNKNOWN = factory.createTypeReferenceNode('unknown')

export const defaultTypeMap = Object.freeze({
  bpchar: 'string',
  char: 'string',
  varchar: 'string',
  text: 'string',
  citext: 'string',
  uuid: 'string',
  bytea: 'string',
  inet: 'string',
  time: 'string',
  timetz: 'string',
  interval: 'string',
  name: 'string',
  int2: 'number',
  int4: 'number',
  int8: 'number',
  float4: 'number',
  float8: 'number',
  numeric: 'number',
  money: 'number',
  oid: 'number',
  bool: 'boolean',
  json: 'unknown',
  jsonb: 'unknown',
  date: 'Date',
  timestamp: 'Date',
  timestamptz: 'Date',
})

export default class TypeRegistry {
  public readonly typeMap: Record<string, string>

  constructor() {
    this.typeMap = { ...defaultTypeMap }
  }

  public has(typename: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.typeMap, typename)
  }

  public get(typename: string): TypeNode {
    if (this.has(typename)) {
      return factory.createTypeReferenceNode(this.typeMap[typename])
    }

    console.warn(`Unknown type detected: '${typename}'. You have either disabled generation of the type or this is a bug.`)
    this.add(typename, 'unknown')

    return UNKNOWN
  }

  public add(typename: string, target: string): void {
    if (this.has(typename)) {
      console.warn(`Re-registering known type '${typename}': ${this.typeMap[typename]} => ${target}`)
    }

    this.typeMap[typename] = target;
  }
}
