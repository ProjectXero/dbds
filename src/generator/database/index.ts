export { default as SchemaInfo } from './SchemaInfo'
export * from './SchemaInfo'
export { default as TypeRegistry } from './TypeRegistry'
