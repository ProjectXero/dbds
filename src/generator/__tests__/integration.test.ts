import assert from 'assert'
import { createPool, DatabasePool, sql } from 'slonik'
import { raw } from 'slonik-sql-tag-raw'
import { Generator } from '..'
import { SchemaInfo } from '../database'

let pool: DatabasePool
let instance: Generator

const SCHEMA = 'generator'

beforeAll(async () => {
  assert(process.env.DATABASE_URL, 'DATABASE_URL must be configured')

  pool = await createPool(process.env.DATABASE_URL, {
    captureStackTrace: true,
    maximumPoolSize: 1,
    idleTimeout: 'DISABLE_TIMEOUT',
    idleInTransactionSessionTimeout: 'DISABLE_TIMEOUT',
  })

  await pool.query(sql.unsafe`CREATE SCHEMA ${sql.identifier([SCHEMA])}`)
  await pool.query(sql.unsafe`SET search_path TO ${raw(SCHEMA)}`)

  await pool.query(sql.unsafe`
    CREATE TYPE "test_type_enum" AS ENUM (
      'aa',
      'bb',
      'zzzz',
      'cc'
    )
  `)
  await pool.query(sql.unsafe`
    CREATE DOMAIN "test_type_domain" AS text
  `)
  await pool.query(sql.unsafe`
    CREATE TABLE "test_table_standard" (
      id SERIAL NOT NULL,
      arr_test TEXT[],
      domain_test TEST_TYPE_DOMAIN,
      enum_test TEST_TYPE_ENUM NOT NULL,
      enum_arr_test TEST_TYPE_ENUM[],
      jsonb_test JSONB,
      "casetest_lower" TEXT,
      "caseTest_upper" TEXT,
      "created_at" TIMESTAMP NOT NULL DEFAULT NOW()
    )
  `)
  await pool.query(sql.unsafe`
    CREATE TABLE "test_table_types" (
      id UUID NOT NULL,
      table_test TEST_TABLE_STANDARD,
      table_arr_test TEST_TABLE_STANDARD[] DEFAULT '{}' NOT NULL
    )
  `)
  await pool.query(sql.unsafe`
    CREATE TABLE "test_table_order" (
      subsequent_table_type TEST_TABLE_STANDARD
    )
  `)
  await pool.query(sql.unsafe`
    CREATE TABLE "test_generation_defaults" (
      generated_id INTEGER NOT NULL GENERATED ALWAYS AS IDENTITY,
      generated_column INTEGER NOT NULL
        GENERATED ALWAYS AS (generated_id + 1) STORED
    )
  `)
})

afterAll(async () => {
  await instance.destroy()
})

beforeEach(() => {
  instance = new Generator({
    schema: new SchemaInfo(pool, SCHEMA),
    genInsertSchemas: true,
    genSelectSchemas: true,
    genTableMetadata: true,
    genEnums: true,
    genInsertTypes: true,
    genSchemaObjects: true,
    genTables: true,
    genTypeObjects: true,
  })
})

describe('Generator', () => {
  it('generates the correct types', async () => {
    expect(await instance.build()).toMatchSnapshot()
  })
})
