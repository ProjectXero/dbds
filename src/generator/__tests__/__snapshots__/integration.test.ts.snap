// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator generates the correct types 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export type MapToSerializable<T> = T extends SerializableValueType ? T : T extends Array<infer U> ? Array<MapToSerializable<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<MapToSerializable<U>> : T extends {
    [K in keyof T]: K extends string ? T[K] extends SimpleValueType ? T[K] : T[K] extends Function ? never : MapToSerializable<T[K]> : never;
} ? T : never;
export enum TestTypeEnum {
    Aa = \\"aa\\",
    Bb = \\"bb\\",
    Zzzz = \\"zzzz\\",
    Cc = \\"cc\\"
}
export interface TestTableOrder {
    subsequent_table_type: TestTableStandard | null;
}
export interface TestTableOrder$Insert {
    subsequent_table_type?: TestTableStandard | null;
}
export const TestTableOrder$Types: Record<keyof TestTableOrder, string> = {
    subsequent_table_type: \\"test_table_standard\\"
};
export interface TestTableStandard<TJsonbTest = SerializableValueType> {
    id: number;
    arr_test: string[] | null;
    domain_test: string | null;
    enum_test: TestTypeEnum;
    enum_arr_test: TestTypeEnum[] | null;
    jsonb_test: MapToSerializable<TJsonbTest> | null;
}
export interface TestTableStandard$Insert<TJsonbTest = SerializableValueType> {
    id?: number;
    arr_test?: string[] | null;
    domain_test?: string | null;
    enum_test: TestTypeEnum;
    enum_arr_test?: TestTypeEnum[] | null;
    jsonb_test?: MapToSerializable<TJsonbTest> | null;
}
export const TestTableStandard$Types: Record<keyof TestTableStandard, string> = {
    id: \\"int4\\",
    arr_test: \\"text\\",
    domain_test: \\"text\\",
    enum_test: \\"test_type_enum\\",
    enum_arr_test: \\"test_type_enum\\",
    jsonb_test: \\"jsonb\\"
};
export interface TestTableTypes {
    id: string;
    table_test: TestTableStandard | null;
    table_arr_test: TestTableStandard[];
}
export interface TestTableTypes$Insert {
    id: string;
    table_test?: TestTableStandard | null;
    table_arr_test?: TestTableStandard[];
}
export const TestTableTypes$Types: Record<keyof TestTableTypes, string> = {
    id: \\"uuid\\",
    table_test: \\"test_table_standard\\",
    table_arr_test: \\"test_table_standard\\"
};
"
`;
